# Example of using the VLSD algorithm

## About this document

The purpose of this document is to show an example of using the VLSD algorithm.

In the following sections, an introduction of the this algorithm will be presented, followed by instructions to install the `subgroups` library. Then, the execution process of the VLSD algorithm will be described, including the necessary steps to consider. Finally, the results obtained from the application of this algorithm will be presented, highlighting the information obtained in the output file and the one that can be accessed through the model properties.

## VLSD algorithm

VLSD (Vertical List Subgroup Discovery) [[1]](#1) is a subgroup discovery algorithm that combines an equivalence class exploration strategy and a pruning strategy based on optimistic estimate, and that is based on the vertical list data structure.

## Installing the `subgroups` library

To install the `subgroups` library, you have to execute the following cell:

In [1]:
!pip install subgroups

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 253.9/253.9 KB 7.7 MB/s eta 0:00:00


After that, to verify that the installation was successful, yo can run the following cell:

In [2]:
import subgroups.tests as st
st.run_all_tests()

test_Operator_evaluate_method (tests.core.test_operator.TestOperator) ... ok
test_Operator_evaluate_method_with_pandasSeries (tests.core.test_operator.TestOperator) ... ok
test_Operator_generate_from_str_method (tests.core.test_operator.TestOperator) ... ok
test_Operator_string_representation (tests.core.test_operator.TestOperator) ... ok
test_Pattern_contains_method (tests.core.test_pattern.TestPattern) ... ok
test_Pattern_general (tests.core.test_pattern.TestPattern) ... ok
test_Pattern_is_contained_method (tests.core.test_pattern.TestPattern) ... ok
test_Pattern_is_refinement_method (tests.core.test_pattern.TestPattern) ... ok
test_Selector_attributes (tests.core.test_selector.TestSelector) ... ok
test_Selector_comparisons (tests.core.test_selector.TestSelector) ... ok
test_Selector_creation_process (tests.core.test_selector.TestSelector) ... ok
test_Selector_deletion_process (tests.core.test_selector.TestSelector) ... ok
test_Selector_generate_from_str_method (tests.core.test_selec



##################################
########## CORE PACKAGE ##########
##################################


##############################################
########## QUALITY MEASURES PACKAGE ##########
##############################################


#############################################
########## DATA STRUCTURES PACKAGE ##########
#############################################


/home/dionnegantzert/.local/lib/python3.10/site-packages/subgroups/data_structures/fp_tree_for_sdmap.py:134: FutureWarning: The provided callable <function sum at 0x7c0baa198790> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  tp_and_fp_for_each_value = target_attribute_as_a_mask.groupby(current_Series).aggregate([size, sum]) # tp -> sum; fp -> size - sum; n -> size.
/home/dionnegantzert/.local/lib/python3.10/site-packages/subgroups/data_structures/fp_tree_for_sdmap.py:134: FutureWarning: The provided callable <function sum at 0x7c0baa198790> is currently using SeriesGroupBy.sum. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "sum" instead.
  tp_and_fp_for_each_value = target_attribute_as_a_mask.groupby(current_Series).aggregate([size, sum]) # tp -> sum; fp -> size - sum; n -> size.
/home/dionne



########################################
########## ALGORITHMS PACKAGE ##########
########################################


ok
test_BSD_cardinality (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_checkRel (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_checkRelevancies (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_fit1 (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_fit2 (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_fit3 (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_fit4 (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_init_method (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_BSD_logicalAnd (tests.algorithms.subgroup_sets.test_bsd.TestBSD) ... ok
test_CBSD_checkRel (tests.algorithms.subgroup_sets.test_cbsd.TestCBSD) ... ok
test_CBSD_checkRelevancies (tests.algorithms.subgroup_sets.test_cbsd.TestCBSD) ... ok
test_CBSD_fit1 (tests.algorithms.subgroup_sets.test_cbsd.TestCBSD) ... ok
test_CBSD_fit2 (tests.algorithms.subgroup_sets.test_cbsd.TestCBSD) ... ok
test_CBSD_



###################################
########## UTILS PACKAGE ##########
###################################


## Running the VLSD algorithm

To run the VLSD algorithm on a dataset, it is necessary to follow these steps:

- Load the dataset into a Pandas `DataFrame` object.
- Set the target, which must be a tuple of the form (column_name, value).
- Select the quality measure and optimistic estimate to use.
- Create the VLSD model with the desired parameters and run it.

The following is an example of running this algorithm on a dataset:

In [3]:
import pandas as pd
from subgroups.quality_measures import WRAcc
from subgroups.quality_measures import WRAccOptimisticEstimate1
from subgroups.algorithms import VLSD

dataset = pd.DataFrame({'att1': ['v3', 'v2', 'v1'], 'att2': ['v1', 'v2', 'v3'], 'att3': ['v2', 'v1', 'v1'], 'class': ['no', 'yes', 'no']})
target = ('class', 'yes')

model = VLSD(quality_measure = WRAcc(), q_minimum_threshold  = -1, optimistic_estimate = WRAccOptimisticEstimate1(), oe_minimum_threshold = -1, sort_criterion_in_s1 = VLSD.SORT_CRITERION_NO_ORDER, sort_criterion_in_other_sizes = VLSD.SORT_CRITERION_NO_ORDER, vertical_lists_implementation = VLSD.VERTICAL_LISTS_WITH_BITSETS, write_results_in_file = True, file_path = "./results.txt")
model.fit(dataset, target)

## Results

Running the following cell, we get the subgroups obtained by the algorithm:

In [4]:
with open("./results.txt", "r") as file:
    for current_line in file:
        print(current_line.strip())

Description: [att1 = 'v1'], Target: class = 'yes' ; Sequence of instances tp = bitarray('000') ; Sequence of instances fp = bitarray('001') ; Quality Measure WRAcc = -0.1111111111111111 ; Optimistic Estimate WRAccOptimisticEstimate1 = 0.0 ; tp = 0 ; fp = 1 ; TP = 1 ; FP = 2
Description: [att1 = 'v2'], Target: class = 'yes' ; Sequence of instances tp = bitarray('010') ; Sequence of instances fp = bitarray('000') ; Quality Measure WRAcc = 0.22222222222222224 ; Optimistic Estimate WRAccOptimisticEstimate1 = 0.6666666666666667 ; tp = 1 ; fp = 0 ; TP = 1 ; FP = 2
Description: [att1 = 'v3'], Target: class = 'yes' ; Sequence of instances tp = bitarray('000') ; Sequence of instances fp = bitarray('100') ; Quality Measure WRAcc = -0.1111111111111111 ; Optimistic Estimate WRAccOptimisticEstimate1 = 0.0 ; tp = 0 ; fp = 1 ; TP = 1 ; FP = 2
Description: [att2 = 'v1'], Target: class = 'yes' ; Sequence of instances tp = bitarray('000') ; Sequence of instances fp = bitarray('100') ; Quality Measure WR

Each of these lines represents a subgroup discovered by the algorithm along with some of its characteristics. If we take the first result as an example, we have the following characteristics:

- The subgroup is described by the pattern `[att1 = 'v1']`.
- The target is the one we defined initially, i.e., `class = 'yes'`.
- The quality of the subgroup is measured by the WRAcc measure, which has a value of -0.1111111111111111
- The optimistic estimate of the subgroup has a value of 0.0
- The values of tp, fp, TP, and FP are as follows: tp = 0 ; fp = 1 ; TP = 1 ; FP = 2.

These results have been verified in the output file of the VLSD algorithm run on a toy dataset.

We can also access different statistics about the result:

In [5]:
print("Selected subgroups: ", model.selected_subgroups) # Number of selected subgroups
print("Unselected subgroups: ", model.unselected_subgroups) # Number of unselected subgroups due to not meeting the minimum quality threshold
print("Visited nodes: ", model.visited_nodes) # Number of nodes (subgroups) visited from the search space

Selected subgroups:  20
Unselected subgroups:  0
Visited nodes:  20


# References

<a id="1">[1]</a> 
Lopez-Martinez-Carrasco, A., Juarez, J.M., Campos, M., Canovas-Segura, B. VLSD - An Efficient Subgroup Discovery Algorithm Based on Equivalence Classes and Optimistic Estimate. Algorithms 2023, 16, 274. https://doi.org/10.3390/a16060274